### DISCUSSION TITLE
where are deepgram servers deployed and how to measure the live streaming latency ?

### DISCUSSION THREAD

**Discussion by midshipman** [deepgram/community#612] [General help] [answered]
would be great to get some insight, as we are doing a real-time conversation app, so latency is the key. thanks

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jkroll-deepgram** [comment]
Hi @midshipman, all of our servers are located in the US. You can refer to our guide [Measuring Streaming Latency](https://developers.deepgram.com/docs/measuring-streaming-latency) which has some suggestions and resources.

**Comment by midshipman** [comment]
Hi Julia

Thanks for the help!

I ran latency_zetbdo.py and got the result below, the average transcription
latency much higher than 563 ms measured in the post. Any idea to improve
this?

Thanks


   - Network latency: 0.238390

   Total latency, test from my macbook in Germany, running latency_zetbdo.py

   * Min latency: 0.017
   * Avg latency: 2.415
   * Max latency: 5.670

   Average transcription latency 2.415 - 0.238 = 2.177

**Reply by jkroll-deepgram** [reply]
Hi @midshipman, can you share the URL parameters you're using, and/or a Deepgram request ID? I may be able to suggest some request-level ways to shave off latency, depending on the features you're using.

Our total customer load varies, and is highest during US business hours, so you may observe our streaming latency also vary depending on time of day and day of week.

Overall though, there aren't many other ways that you can control latency. I'd just recommend running tests on different days/times to get a sense of the average latency you're seeing, rather than one isolated test, since I would expect variance.

**Reply by midshipman** [reply]
@jkroll-deepgram fyi, the [original code](url) seems outdated, so I made some changes to run it, comment out check for request_id and is_final, and check for msg type.

The request url is wss://api.deepgram.com/v1/listen?channels={channels}&sample_rate={sample_rate}&encoding=linear16'

async def receiver(ws):
            """ Print out the messages received from the server.
            """
            nonlocal audio_cursor
            transcript_cursor = 0.
            min_latency = 0
            max_latency = 0
            avg_latency_num = 0
            avg_latency_den = 0
            async for msg in ws:
                msg = json.loads(msg)
                # print(msg)
                # if 'request_id' in msg:
                #     # This is the final metadata message. It gets sent as the
                #     # very last message by Deepgram during a clean shutdown.
                #     # There is no transcript in it.
                #   continue
                # if msg['is_final']:
                #      continue
                # Just a moment ago -- right before we received this message --
                # is when the latency was at its worse. So measure max latency
                # and then update the transcript_cursor
                
                # print(f'Measuring... Audio cursor = {audio_cursor:.3f}, Transcript cursor = {transcript_cursor:.3f}')
                # print(msg)
                if(msg['type'] == 'Results'):


                    cur_max_latency = audio_cursor - transcript_cursor

                    transcript_cursor = msg['start'] + msg['duration']

                    # Since we just received a message, latency is currently at its
                    # best.
                    cur_min_latency = audio_cursor - transcript_cursor

                    # The average latency (as would be measured by a constant probe
                    # or by a sampling profiler) is mathematically equivalent to a
                    # weighted sum.
                    avg_latency_num += (cur_min_latency + cur_max_latency) / 2 * msg['duration']
                    avg_latency_den += msg['duration']

                    # Update global max/min latencies.
                    max_latency = max(max_latency or 0, cur_max_latency)
                    min_latency = min(min_latency or 10**6, cur_min_latency)

                    print(f'Measuring... Audio cursor = {audio_cursor:.3f}, Transcript cursor = {transcript_cursor:.3f}')

            print(f'Min latency: {min_latency:.3f}')
            print(f'Avg latency: {avg_latency_num / (avg_latency_den or 1):.3f}')
            print(f'Max latency: {max_latency:.3f}')

**Comment by midshipman** [comment]
thanks and pls find info below:

req id 5779b191-187b-4369-90fc-2ed6dc30cae9

Min latency: 0.017
Avg latency: 2.500
Max latency: 6.030

**Comment by midshipman** [comment]
hi deepgram, can someone help me here? @jkroll-deepgram

**Reply by jkroll-deepgram** [reply, marked as answer]
Hi @midshipman, sorry about that code not working out of the box, and I appreciate you taking time to debug it.

Note that we recommend only measuring latency on interim results, as there is different logic for when final results are provided, which has other dependencies on audio context. See our note in [How do I measure latency with interim results?](https://developers.deepgram.com/docs/interim-results#how-do-i-measure-latency-with-interim-results):
> However, remember that to give you best accuracy, final transcripts may end early (see lines 9 and 10 in the example above), which means you’ve already received more data than what is reflected in the final transcript. Remember that final transcripts are meant for conservative pipelines that require the highest confidence levels, whereas the latest interim transcript has the lowest latency; always ignore final transcripts when calculating latency.

I see that our Python code example has the significant oversight of not including the parameter `interim_results=true` in the Deepgram URL.

If you download the code sample fresh as-is, and then add that parameter to line 43:
```
f'wss://api.deepgram.com/v1/listen?channels={channels}&sample_rate={sample_rate}&encoding=linear16&interim_results=true',
```

Then you can rerun the code and see the true latency metrics.

This is the output that I get:
```
% python3 latency_zetbdo.py -k <API-key> <test-file>.wav
Channels = 1, Sample Rate = 48000 Hz, Sample width = 2 bytes, Size = 1341312 bytes
Measuring... Audio cursor = 1.120, Transcript cursor = 1.000
Measuring... Audio cursor = 2.100, Transcript cursor = 2.000
Measuring... Audio cursor = 3.100, Transcript cursor = 3.000
Measuring... Audio cursor = 4.100, Transcript cursor = 4.000
Measuring... Audio cursor = 5.120, Transcript cursor = 5.000
Measuring... Audio cursor = 6.080, Transcript cursor = 6.000
Measuring... Audio cursor = 7.080, Transcript cursor = 7.000
Measuring... Audio cursor = 8.120, Transcript cursor = 8.020
Measuring... Audio cursor = 9.020, Transcript cursor = 8.860
Measuring... Audio cursor = 9.940, Transcript cursor = 9.860
Measuring... Audio cursor = 11.140, Transcript cursor = 11.060
Measuring... Audio cursor = 12.180, Transcript cursor = 12.060
Measuring... Audio cursor = 13.160, Transcript cursor = 13.060
Min latency: 0.080
Avg latency: 0.600
Max latency: 1.280
```

So in this case, I am seeing an average latency of 600ms, with a range from 80-1280ms.

Note that when I made the change that you did (commenting out the if-statement of `# if msg['is_final']:`, which runs latency metrics on final results, which is not recommended), then the script yields:
```
% python3 latency_zetbdo.py -k <API-key> <test-file>.wav
Channels = 1, Sample Rate = 48000 Hz, Sample width = 2 bytes, Size = 1341312 bytes
Measuring... Audio cursor = 5.180, Transcript cursor = 3.860
Measuring... Audio cursor = 9.060, Transcript cursor = 6.940
Measuring... Audio cursor = 10.200, Transcript cursor = 10.060
Measuring... Audio cursor = 13.580, Transcript cursor = 13.470
Measuring... Audio cursor = 13.980, Transcript cursor = 13.972
Min latency: 0.008
Avg latency: 2.537
Max latency: 5.200
```

**Comment by midshipman** [comment]
Thank you for the clarification. interim_results=true is the key!
